Feature/task defaults

jlariono opened the pull request:
- Added support for setting task parameter default values using a `TaskDefaults` global property in a Whiskey configuration file.
- Added `Add-WhiskeyTaskDefault` function for adding new task default parameter values to an existing task context object.
- Misc test fixes and improvements

jlariono commented:
I'll submit changes after I can rebase my work on top of the changes in PR #132, since that PR affects some of the files in this PR.

jlariono commented:
PR ready for review now. Task default functionality moved into a `TaskDefaults` task.

splatteredbits commented on Whiskey/Functions/Add-WhiskeyTaskDefault.ps1:
```diff
@@ -0,0 +1,75 @@
+
+    if ($Task -notin (Get-WhiskeyTask | Select-Object -ExpandProperty 'Name'))
+    {
+        throw 'The Task ''{0}'' is not a valid Whiskey task.' -f $Task
```
Change error message to `Task ''{0}'' does not exist.`

splatteredbits commented on Whiskey/Functions/New-WhiskeyContext.ps1:
```diff
@@ -8,19 +8,28 @@ function New-WhiskeyContext
     * `ByBuildServer`: a flag indicating if the build is being run by a build server.
     * `ByDeveloper`: a flag indicating if the build is being run by a developer.
-    * `ApplicatoinName`: the name of the application being built.
+    * `BuildMetadata`: metadata information of the current build on a build server.
```
Some of these properties are private and shouldn't be used outside of Whiskey. Remove:

* Configuration
* Credentials
* DownloadRoot
* Task Defaults
* Variables

splatteredbits commented on Test/Invoke-WhiskeyTask.Tests.ps1:
```diff
@@ -584,36 +617,30 @@ function WhenRunningTask
     if( $InRunMode )
     {
-        $context.RunMode = $InRunMode;
+        $script:context.RunMode = $InRunMode
```
You only need the `$script:` scope when setting the value of the variable, not properties on the variable.

splatteredbits commented on Test/Invoke-WhiskeyTask.Tests.ps1:
```diff
@@ -3,6 +3,8 @@ Set-StrictMode -Version 'Latest'
 
```
I don't understand why you made any changes to this test. You only changed one line in Invoke-WhiskeyTask that shouldn't have had any affect on tests. Can you give me some context?

jlariono replied:
When I was making `TaskDefaults` a global config I added some new tests to this file at which point I also did some refactoring to make the tests cleaner. I ended up not needing the tests that I added, but left the refactoring. Should I revert these changes?

splatteredbits replied:
Yes, please.

splatteredbits commented on Test/WhiskeyTest.psm1:
```diff
@@ -294,12 +290,9 @@ function New-WhiskeyTestContext
     }
 
-    $context = New-WhiskeyContextObject
+    $context = New-WhiskeyContext -Environment 'Verification' -ConfigurationPath $ConfigurationPath
```
Why this change? It requires any test that calls it to setup an entire build environment, which is expensive and makes tests have to know everything about Whiskey instead of just what they care about.

jlariono replied:
Oh this change was originally made when I initially made `TaskDefaults` a global config property that was set in `New-WhiskeyContext`. Now that `TaskDefaults` is a task, this is no longer necessary. I'll revert this.

jlariono commented:
@splatteredbits Requested changes completed. Ready for re-review.

jlariono commented:
@splatteredbits Realized a Whiskey release has occurred since I created this branch. I rebased onto develop, incremented the version number, and removed the old release notes. I'll need another approval.

splatteredbits commented:
@jlariono OK. Let me know when you're ready.

jlariono commented:
@splatteredbits It's ready for review now. I rebased onto develop and the only file changed was `Whiskey.psd1`. I incremented the `Whiskey.psd1` version number when resolving the merge conflicts during the rebase.

splatteredbits commented:
@jlariono I didn't see any new commits so couldn't easily see what changed.
